### PR TITLE
Expectation tests now check the encoded AST

### PR DESCRIPTION
**Pull request by sfc-gh-azwiegincew**
The unparser is invoked only if it's specified. We expect to not invoke it during merge gate testing.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/ast.py [hunk 1]
```diff
@@ -121,6 +121,9 @@ def decode_ast_response_from_snowpark(res: dict, session_parameters: Any) -> Any
 
 
 class AstBatch:
+    # Function used to generate request IDs. This is overridden in some tests.
+    request_id_gen_fn = uuid.uuid4
```
Comment (sfc-gh-oplaton) [inline src/snowflake/snowpark/_internal/ast.py:132]: Trivial topic: I know fields are usually nouns, but in true OOP (like in Python), there's no substantial difference between fields and methods anyway. Should this name simply be `generate_request_id`?

File: tests/ast/data/RelationalGroupedDataFrame.test [hunk 2]
```diff
@@ -33,7 +33,9 @@ df.group_by("empid").pivot("month", ['JAN', 'FEB']).sum("amount").show()
 
 df.group_by(["empid", "team"]).pivot("month").sum("amount").sort("empid", "team").show()
 
-## EXPECTED OUTPUT
+## EXPECTED ENCODED AST
+
```
Comment (sfc-gh-oplaton) [inline tests/ast/data/RelationalGroupedDataFrame.test:37]: Is this working as intended?

File: tests/ast/data/df_random_split.test [hunk 3]
```diff
@@ -8,7 +8,9 @@ df2 = df.random_split(weights)
 
 df3 = df.random_split(weights, seed=24)
 
-## EXPECTED OUTPUT
+## EXPECTED ENCODED AST
+
```
Comment (sfc-gh-oplaton) [inline tests/ast/data/df_random_split.test:12]: Is this working as intended?

File: tests/ast/data/functions.test [hunk 4]
```diff
@@ -594,7 +594,9 @@ df303 = df.select(unix_timestamp("A"), unix_timestamp("A", None), unix_timestamp
 
 df304 = df.select(locate("needle", col("expr")), locate("needle", lit("test string"), 2))
 
-## EXPECTED OUTPUT
+## EXPECTED ENCODED AST
+
```
Comment (sfc-gh-oplaton) [inline tests/ast/data/functions.test:598]: Is this working as intended?

File: tests/ast/data/session_write_pandas.test [hunk 5]
```diff
@@ -8,7 +8,9 @@ ans = session.write_pandas(df, "test_table")
 
 ans2 = session.write_pandas(df, "test", schema="a", database="b", chunk_size=7, compression="brotli", on_error="ignore", parallel=10, quote_identifiers=False, auto_create_table=True, create_temp_table=True, overwrite=True, table_type="temp", random=90)
 
-## EXPECTED OUTPUT
+## EXPECTED ENCODED AST
+
```
Comment (sfc-gh-oplaton) [inline tests/ast/data/session_write_pandas.test:12]: Is this working as intended?

File: tests/ast/test_ast_driver.py [hunk 6]
```diff
@@ -219,17 +241,21 @@ def test_ast(session, test_case):
...
     else:
         try:
-            assert actual.strip() == test_case.expected_output.strip()
+            assert base64.strip() == test_case.expected_ast_base64.strip()
+            if pytest.unparser_jar:
+                assert actual.strip() == test_case.expected_ast_unparsed.strip()
         except AssertionError as e:
             raise AssertionError(
-                f"If the expectation is incorrect, run pytest --update-expectations:\n\n{base64}\n{e}"
+                f"If the expectation is incorrect, run pytest --update-expectations:\n{e}\nFUFU\n\n\n{base64} FUFU \n\n\n{test_case.expected_ast_base64}"
```
Comment (sfc-gh-oplaton) [inline tests/ast/test_ast_driver.py:258]: I'm guessing the message needs an update before merging?

File: tests/ast/data/RelationalGroupedDataFrame.test (hunk 2 above)
Comment (sfc-gh-azwiegincew) [reply tests/ast/data/RelationalGroupedDataFrame.test:37]: yep

File: tests/ast/README.md [hunk 7]
```diff
@@ -12,7 +12,11 @@ N.B. No eager evaluation is permitted, as any intermediate batches will not be o
 df = session.table("test_table")
 df = df.filter("STR LIKE '%e%'")
 
-## EXPECTED OUTPUT
+## EXPECTED ENCODED AST
```
Comment (sfc-gh-lspiegelberg) [inline tests/ast/README.md:15]: could we maybe flip the order here? i.e. EXPECTED ENCODED AST as the last section?

I think it's much easier to read to have the base64 strings at the end.

File: tests/ast/test_ast_driver.py [hunk 8]
```diff
@@ -110,8 +119,20 @@ def run_test(session, test_name, test_source):
...
+
+import uuid
+
+# Set up the request ID generator.
+AstBatch.generate_request_id = lambda: uuid.uuid5(uuid.NAMESPACE_DNS, "id-gen")
+
+ast_utils.SRC_POSITION_TEST_MODE = True
 
 def run_test(session):
+    # Reset the entity ID generator.
```
Comment (sfc-gh-lspiegelberg) [inline tests/ast/test_ast_driver.py:133]: nit: this should be a public function (or private `_reset`). Calling constructor manually sounds ... dangerous?

**Review by sfc-gh-lspiegelberg — APPROVED** [review]
lgtm! big improvement. I really like that this also shows how large/complex the ASTs are.

**Pull request merged**